File: model/src/main/java/ru/mipt/acsl/decode/model/domain/DecodeName.java
package ru.mipt.acsl.decode.model.domain;

import com.google.common.base.Preconditions;
import org.jetbrains.annotations.NotNull;

/**
 * @author Artem Shein
 */
public interface DecodeName extends DecodeElement
{
    @NotNull
    String asString();

    @NotNull
    static String mangleName(@NotNull String name)
    {
        if (name.startsWith("^"))
        {
            name = name.substring(1);
        }
        name = name.replaceAll("[ \\\\^]", "");
        Preconditions.checkState(!name.isEmpty(), "invalid name");
        return name;
    }
}

File: model/src/main/scala/ru/mipt/acsl/decode/model/domain/impl/DecodeName.scala
package ru.mipt.acsl.decode.model.domain.impl

import ru.mipt.acsl.decode.model.domain

/**
  * @author Artem Shein
  */
case class DecodeName(value: String) extends domain.IDecodeName {
  override def asString(): String = value
}

object DecodeName {
  def newFromSourceName(name: String) = DecodeName(domain.IDecodeName.mangleName(name))
  def newFromMangledName(name: String) = DecodeName(name)
}
